src/Misc/coroutine.ts

```diff
@@ -0,0 +1,133 @@
+import { Deferred } from './deferred';
+
+// "Coroutines are computer program components that generalize subroutines for non-preemptive multitasking, by allowing execution to be suspended and resumed."
+// https://en.wikipedia.org/wiki/Coroutine
+
+// In this implementation, coroutines are typically created via generator functions (function* with yield statements).
+// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/function*
+
+// In this implementation, the generator function (possibly parameterized) is referred to as a coroutine factory, and the returned iterator is referred to as the coroutine.
+// Technically yielding generator functions are not required - anything that implements the contract of Coroutine<T> can be run as a coroutine.
+
+// The coroutine is started with the first call to next on the iterator, it is suspended with yield statements, and it is resumed with additional calls to next on the iterator.
+// To create an object satisfying the Coroutine<T> contract with a generator function, it must not yield values, but rather only undefined via a plain "yield;" statement.
+// Coroutines can call other coroutines via:
+// 1. yield* someOtherCoroutine(); // If the called coroutine does not return a value
+// 2. const result = yield* someOtherCoroutine(); // If the called coroutine returns a value
+
+// Coroutines are run with the runCoroutine function, which takes a Coroutine<T>, a CoroutineScheduler<T>, and a success and error callback.
+// A scheduler is responsible for scheduling the next step of a coroutine, either synchronously or asynchronously.
+
+/**
+ * A Coroutine<T> is the intersection of:
+ * 1. An Iterator that yields undefined, returns a T, and is not passed values with calls to next.
+ * 2. An IterableIterator of undefined (since it only yields undefined).
+ */
+export type Coroutine<T> = Iterator<undefined, T, unknown> & IterableIterator<undefined>;
```

Comment:
Is this meaningfully different from the coroutines we already have?

https://github.com/BabylonJS/Babylon.js/blob/master/src/Misc/observableCoroutine.ts
https://doc.babylonjs.com/divingDeeper/events/coroutines

Reply:
Never mind, just saw this had already been noted. Will sync separately.